### DISCUSSION TITLE
Caching of same but differently loaded dataset

### DISCUSSION THREAD

**Discussion by Revist** [huggingface/datasets#6593] [Q&A]
I have two datasets: COCO and COCO-20i. These datasets are the same but they are loaded in a different way i.e. one returns
image + annorations for this image, second returns 6 images with a given class + annotations of this class. The underlying data _ anntations are the same, the only difference is how _generate_examples is executed. I could create dataset loading script with these two configurations. 

My questions is:
As this results in arrow files in the cache will these replicate the image data i.e. for every _generate_examples i have different dataset so different arrow files resulting in dataset growing with number of configurations?

**Comment by dexhunter** [comment]
Different builder configurations produce separate prepared Arrow datasets, so the generated rows and annotations are stored separately. Whether that duplicates the **image bytes** depends on how the image feature is encoded.

- If examples contain paths to local image files, an `Image` feature can store the path with `bytes=None`. Both configurations can refer to the same downloaded/extracted files; their Arrow tables repeat paths and annotations, not the image contents.
- If examples contain encoded bytes or PIL images without a filename, those bytes are embedded in Arrow. Reusing the same image in several generated rows can then duplicate its payload; the dataset does not automatically deduplicate that content across configurations.

You can inspect the representation without decoding an image:

```python
from datasets import Image

raw = ds.cast_column("image", Image(decode=False))
print(raw[0]["image"])  # {"path": ..., "bytes": ...}
```

For your two views, keep the image files in one stable location and generate references to them if avoiding duplicate image storage is the priority. The six-image view can likewise contain a list of image references. Those files must remain available when the dataset is read.

The distinction is implemented in [`Image.encode_example`](https://github.com/huggingface/datasets/blob/19f69de53015525126e6d3f96859acb90498e364/src/datasets/features/image.py). Sharing via `push_to_hub()` is a separate step and embeds external image files by default.
